cpp/src/arrow/parquet/writer.cc

```diff
@@ -143,6 +153,82 @@ Status FileWriter::Close() {
 
 FileWriter::~FileWriter() {}
 
+// Create a slice of a PrimitiveArray.
+//
+// This method is specially crafted for WriteFlatTable and assumes the following:
+//  * chunk_size is a multiple of 512
+Status TemporaryArraySlice(int64_t chunk, int64_t chunk_size, const PrimitiveArray* array,
+    std::shared_ptr<PrimitiveArray>* out) {
+  // The last chunk may be smaller than the chunk_size
+  const int64_t size = std::min(chunk_size, array->length() - chunk * chunk_size);
+  const int64_t buffer_offset = chunk * chunk_size * array->type()->value_size();
+  const int64_t value_size = size * array->type()->value_size();
+  auto chunk_buffer = std::make_shared<Buffer>(array->data(), buffer_offset, value_size);
```

Comment:
I've been meaning to send out an e-mail to the list about this (already opened up a JIRA to discuss).  But technically we should wrap this in a try/catch block to catch std::alloc_error.  Ideally we would have a utility function that wraps make shared that returns Status, and assigns the shared pointer as an output variable.